langchain/callbacks/flyte_callback.py

```diff
@@ -0,0 +1,393 @@
+    ) -> None:
+        """Run when LLM starts."""
+
+        self.metrics["step"] += 1
+        self.metrics["llm_starts"] += 1
+        self.metrics["starts"] += 1
+
+        resp: Dict[str, Any] = {}
+        resp.update({"action": "on_llm_start"})
+        resp.update(flatten_dict(serialized))
+        resp.update(self.metrics)
+
+        prompt_responses = []
+        for prompt in prompts:
+            prompt_responses.append(prompt)
+
+        resp.update({"prompts": prompt_responses})
+
+        self.deck.append(self.markdown_renderer().to_html("### LLM Start"))
+        self.deck.append(
+            self.table_renderer().to_html(self.pandas.DataFrame([resp])) + "\n"
+        )
+
+    def on_llm_new_token(self, token: str, **kwargs: Any) -> None:
+        """Run when LLM generates a new token."""
+        pass
+
+    def on_llm_end(self, response: LLMResult, **kwargs: Any) -> None:
+        """Run when LLM ends running."""
+        self.metrics["step"] += 1
+        self.metrics["llm_ends"] += 1
+        self.metrics["ends"] += 1
+
+        resp: Dict[str, Any] = {}
+        resp.update({"action": "on_llm_end"})
+        resp.update(flatten_dict(response.llm_output or {}))
+        resp.update(self.metrics)
+
+        all_complexity_metrics = []
+
+        for generations in response.generations:
+            for generation in generations:
+                generation_resp = deepcopy(resp)
+                generation_resp.update(flatten_dict(generation.dict()))
+                generation_resp.update(
+                    analyze_text(
+                        generation.text,
+                        nlp=self.nlp,
+                    )
+                )
+                complexity_metrics: Dict[str, float] = generation_resp.pop("text_complexity_metrics")  # type: ignore  # noqa: E501
+                all_complexity_metrics.append(complexity_metrics)
+
+                dependency_tree = generation_resp["dependency_tree"]
+
+        self.deck.append(self.markdown_renderer().to_html("### LLM End"))
+        self.deck.append(self.table_renderer().to_html(self.pandas.DataFrame([resp])))
+        self.deck.append(
+            self.markdown_renderer().to_html("#### Text Complexity Metrics")
+        )
+        self.deck.append(
+            self.table_renderer().to_html(
+                self.pandas.DataFrame.from_records(all_complexity_metrics)
+            )
+        )
+        self.deck.append(self.markdown_renderer().to_html("#### Dependency Tree"))
+        self.deck.append(dependency_tree)
+
+    def on_llm_error(
+        self, error: Union[Exception, KeyboardInterrupt], **kwargs: Any
+    ) -> None:
+        """Run when LLM errors."""
+        self.metrics["step"] += 1
+        self.metrics["errors"] += 1
+
+    def on_chain_start(
+        self, serialized: Dict[str, Any], inputs: Dict[str, Any], **kwargs: Any
+    ) -> None:
+        """Run when chain starts running."""
+        import pandas as pd
```

Comment:
same pandas import, if pandas is not available this will fail

Reply:
It will fail on initialization through the lazy import so this is OK, but if we're already assigningt he package as an instance attribute then we can just access that

Reply:
can we use `self.pandas` here?

Reply:
@cosmicBboy, good catch! Removed the imports and using instance attributes now.